File: backend/internal/logger/logger_test.go
package logger

import (
	"log/slog"
	"testing"
)

func TestParseLevel(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want slog.Level
	}{
		{name: "debug", in: "debug", want: slog.LevelDebug},
		{name: "info", in: "info", want: slog.LevelInfo},
		{name: "warn", in: "warn", want: slog.LevelWarn},
		{name: "warning も warn 扱い", in: "warning", want: slog.LevelWarn},
		{name: "error", in: "error", want: slog.LevelError},
		{name: "大文字小文字を無視", in: "DEBUG", want: slog.LevelDebug},
		{name: "前後の空白を無視", in: "  info  ", want: slog.LevelInfo},
		{name: "空文字は既定の info", in: "", want: slog.LevelInfo},
		{name: "未知の値は既定の info", in: "verbose", want: slog.LevelInfo},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := parseLevel(tt.in); got != tt.want {
				t.Errorf("parseLevel(%q) = %v, want %v", tt.in, got, tt.want)
			}
		})
	}
}

File: web/src/lib/logger.test.ts
import { describe, expect, it } from "vitest";
import { shouldLog, thresholdFor } from "./logger";

describe("thresholdFor", () => {
  it("開発(PROD=false)なら debug まで出す", () => {
    expect(thresholdFor({ PROD: false })).toBe("debug");
  });

  it("本番(PROD=true)なら warn 以上に絞る", () => {
    expect(thresholdFor({ PROD: true })).toBe("warn");
  });

  it("VITE_LOG_LEVEL が妥当なら本番でも優先する", () => {
    expect(thresholdFor({ PROD: true, VITE_LOG_LEVEL: "debug" })).toBe("debug");
  });

  it("VITE_LOG_LEVEL が未知の値なら無視して既定にフォールバック", () => {
    expect(thresholdFor({ PROD: false, VITE_LOG_LEVEL: "verbose" })).toBe(
      "debug",
    );
  });
});

describe("shouldLog", () => {
  it("閾値と同じレベルは出す", () => {
    expect(shouldLog("warn", "warn")).toBe(true);
  });

  it("閾値より深刻なら出す", () => {
    expect(shouldLog("error", "warn")).toBe(true);
  });

  it("閾値より軽いレベルは出さない", () => {
    expect(shouldLog("info", "warn")).toBe(false);
    expect(shouldLog("debug", "info")).toBe(false);
  });
});
